File: src-tauri/src/commands/books.rs
use crate::error::AppError;
use crate::models::CreateBookRequest;
use chrono::Utc;
use uuid::Uuid;

#[tauri::command]
pub fn generate_book_id() -> String {
    Uuid::new_v4().to_string()
}

#[tauri::command]
pub fn get_current_timestamp() -> String {
    Utc::now().to_rfc3339()
}

#[tauri::command]
pub fn validate_book_request(request: CreateBookRequest) -> Result<(), AppError> {
    if request.title.trim().is_empty() {
        return Err(AppError::from("Title cannot be empty"));
    }
    if request.total_progress <= 0 {
        return Err(AppError::from("Total progress must be greater than 0"));
    }
    let valid_types = ["page", "location", "percentage"];
    if !valid_types.contains(&request.progress_type.as_str()) {
        return Err(AppError::from(
            "Progress type must be 'page', 'location', or 'percentage'",
        ));
    }
    Ok(())
}

#[tauri::command]
pub fn validate_progress_update(
    current: i32,
    total: i32,
    progress_type: String,
) -> Result<(), AppError> {
    if current < 0 {
        return Err(AppError::from("Progress cannot be negative"));
    }
    if current > total {
        return Err(AppError::from("Progress cannot exceed total"));
    }
    Ok(())
}

File: wails/service/books.go
package service

import (
	"errors"
	"strings"

	"github.com/google/uuid"
	"github.com/maxachis/book-tracker/wails/model"
)

// GenerateBookID returns a fresh RFC 4122 v4 UUID for use as a book primary key.
func GenerateBookID() string {
	return uuid.NewString()
}

// ValidateCreateBookRequest mirrors the Rust validate_book_request check:
// non-empty title, positive total_progress, and a recognised progress_type.
func ValidateCreateBookRequest(r model.CreateBookRequest) error {
	if strings.TrimSpace(r.Title) == "" {
		return errors.New("title cannot be empty")
	}
	if r.TotalProgress <= 0 {
		return errors.New("total progress must be greater than 0")
	}
	switch r.ProgressType {
	case string(model.ProgressPage), string(model.ProgressLocation), string(model.ProgressPercentage):
		return nil
	default:
		return errors.New("progress type must be 'page', 'location', or 'percentage'")
	}
}
